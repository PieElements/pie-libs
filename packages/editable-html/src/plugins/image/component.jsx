import { Data, Raw } from 'slate';

import { LinearProgress } from 'material-ui/Progress';
import Portal from 'react-portal';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import debug from 'debug';
import { primary } from '../../theme';
import { withStyles } from 'material-ui/styles';

const log = debug('editable-html:plugins:image:component');

const logError = debug('editable-html:plugins:image:component');

logError.log = console.error.bind(console);

export class RawComponent extends React.Component {

  getWidth = (percent) => {
    const multiplier = percent / 100;
    return this.img.naturalWidth * multiplier;
  }

  getHeight = (percent) => {
    const multiplier = percent / 100;
    return this.img.naturalHeight * multiplier;
  }

  getPercentFromWidth = (width) => {
    var floored = (width / this.img.naturalWidth) * 4;
    return parseInt(floored.toFixed(0) * 25, 10);
  }

  applySizeData = () => {
    const { node, editor } = this.props;

    const resizePercent = node.data.get('resizePercent');

    let update = node.data;

    if (resizePercent) {
      const w = update.get('width');
      update = update.set('width', this.getWidth(resizePercent));

      const h = update.get('height');
      update = update.set('height', this.getHeight(resizePercent));
    } else {
      const w = update.get('width');
      if (w) {
        update = update.set('resizePercent', this.getPercentFromWidth(w));
      }
    }

    if (!update.equals(node.data)) {
      editor.change(c => c.setNodeByKey(node.key, { data: update }));
    }
  }

  componentDidMount() {
    this.applySizeData();
  }

  componentDidUpdate() {
    this.applySizeData();
  }

  render() {

    const { node, editor, classes, attributes, onFocus, onBlur } = this.props;
    const active = editor.value.isFocused && editor.value.selection.hasEdgeIn(node);
    const src = node.data.get('src');
    const percent = node.data.get('percent');
    const resizePercent = node.data.get('resizePercent');
    const loaded = node.data.get('loaded') !== false;
    const deleteStatus = node.data.get('deleteStatus');

    const style = {
      width: resizePercent ? `${node.data.get('width')}px` : 'auto',
      height: resizePercent ? `${node.data.get('height')}px` : 'auto'
    }

    const className = classNames(
      classes.root,
      active && classes.active,
      !loaded && classes.loading,
      deleteStatus === 'pending' && classes.pendingDelete);

    const progressClasses = classNames(
      classes.progress,
      loaded && classes.hideProgress);

    const resize = (amount) => this.resizeBy.bind(this, amount);
    log('showDelete: loaded: ', loaded, 'deleteStatus: ', deleteStatus);

    const showDelete = editor.value.isFocused && loaded && deleteStatus !== 'pending';

    return <div
      onFocus={onFocus}
      className={className}>
      <LinearProgress
        mode="determinate"
        value={percent > 0 ? percent : 0}
        className={progressClasses} />
      <img
        src={src}
        {...attributes}
        ref={r => this.img = r}
        style={style} />
    </div>
  }
}

RawComponent.propTypes = {
  onDelete: PropTypes.func.isRequired
}

const styles = {
  portal: {
    position: 'absolute',
    opacity: 0,
    transition: 'opacity 200ms linear'
  },
  floatingButtonRow: {
    backgroundColor: 'white',
    borderRadius: '1px',
    display: 'flex',
    padding: '10px',
    border: 'solid 1px #eeeeee',
    boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)'
  },
  progress: {
    position: 'absolute',
    left: '0',
    width: '100%',
    top: '0%',
    transition: 'opacity 200ms linear'
  },
  hideProgress: {
    opacity: 0
  },
  loading: {
    opacity: 0.3
  },
  pendingDelete: {
    opacity: 0.3
  },
  root: {
    position: 'relative',
    border: 'solid 1px white',
    display: 'inline-flex',
    transition: 'opacity 200ms linear'
  },
  active: {
    border: `solid 1px ${primary}`
  },
  delete: {
    position: 'absolute',
    right: 0,
  }
}

export default withStyles(styles)(RawComponent);