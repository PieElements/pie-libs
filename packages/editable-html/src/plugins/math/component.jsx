import { Data, findDOMNode } from 'slate';
import { addBrackets, removeBrackets } from '@pie-libs/math-input';

import MathWrapper from './input-wrapper';
import React from 'react';
import classNames from 'classnames';
import debug from 'debug';
import injectSheet from 'react-jss';
import { primary } from '../../theme';

const log = debug('editable-html:plugins:math:component');

export class MathComponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      disableBlur: false
    }

    this.onFocus = (event) => {
      log('[onFocus]')
      if (this.props.onFocus) {
        this.props.onFocus(event);
      }
    }

    this.onBlur = (event) => {
      log('[onBlur]')
      if (this.props.onBlur) {
        this.props.onBlur(event);
      }
    }

    this.onChange = (latex) => {
      log('[onChange]', latex);
      const { node, editor } = this.props;
      const { key } = node;
      const data = Data.create({ latex, editing: true });
      const change = editor.value.change().setNodeByKey(key, { data });
      editor.onChange(change);
    }

    this.onClick = (event) => {
      log('[onClick] preventDefault and stopPropagation');
      if (this.props.onClick) {
        event.preventDefault();
        event.stopPropagation();
        this.props.onClick();
      }
    }

    this.onDeleteClick = (event) => {
      log('delete click');
      event.preventDefault();
      event.stopPropagation();

      const { node, editor } = this.props;

      const change = editor.value.change().moveNodeByKey(node.key);
      editor.onChange(change);
    }
  }

  componentDidUpdate() {
    log('this.wrapper: ', this.wrapper)
    const { node, editor } = this.props;
    const mathChange = node.data.get('change');

    if (mathChange) {
      const latex = this.wrapper.change(mathChange);

      const data = node.data.toObject();
      delete data.change;
      data.latex = latex;

      log('[componentDidUpdate] new latex: ', data.latex);

      const change = editor.value.change().setNodeByKey(node.key, { data })
      editor.change(c => {
        c.setNodeByKey(node.key, { data })
      });
    }
  }

  componentDidMount() {
    log('this.wrapper: ', this.wrapper)
  }

  shouldComponentUpdate(nextProps) {
    const { node: nextNode } = nextProps;
    const { node } = this.props;
    if (nextNode.data.equals(node.data)) {
      return false;
    }

    if (nextNode.data.get('change') === undefined && node.data.get('change') !== undefined) {
      return false;
    }
    return true;
  }

  onMathChange = (latex) => {
    log('[onMathChange]', latex);
    const { node, editor } = this.props;
    const data = Data.create({ latex });
    editor.change(c => c.setNodeByKey(node.key, { data }));
  }

  blur() {
    log('[blur]')
    this.wrapper && this.wrapper.blur();
  }

  render() {
    const { node, classes, editor } = this.props;
    log('[render] >> node', node, 'editor: ', editor, 'value: ', editor.value, editor.value.isFocused);

    const latex = node.data.get('latex');
    const names = classNames(classes.root);
    const cleanLatex = removeBrackets(latex);

    return (
      <div className={names}>
        <MathWrapper
          ref={r => this.wrapper = r}
          latex={cleanLatex}
          editing={true}
          onClick={this.onClick}
          onChange={this.onMathChange}
          onFocus={this.onFocus}
          onBlur={this.onBlur} />
      </div>
    );
  }
}

const styles = {
  root: {
    display: 'inline-flex',
    alignItems: 'center',
    '& > .mq-editable-field': {

      border: 'solid 1px lightgrey',
    },
    '& > .mq-focused': {
      outline: 'none',
      boxShadow: 'none',
      border: `solid 1px ${primary}`,
      borderRadius: '0px'
    }
  },
  selected: {
    border: 'solid 1px red'
  }
};

export default injectSheet(styles)(MathComponent);