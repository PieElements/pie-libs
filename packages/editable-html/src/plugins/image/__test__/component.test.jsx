import { Data } from 'slate';
import { RawComponent } from '../component';
import React from 'react';
import renderer from 'react-test-renderer';
import { stub } from 'sinon';

it('renders correctly', () => {

  const node = {
    data: Data.create({
      resizePercent: 50,
      width: 50,
      height: 50,
      percent: 50
    })
  };

  const editor = {
    value: {},
    change: stub()
  }

  const onDelete = stub();

  const classes = {
    active: 'active',
    loading: 'loading',
    pendingDelete: 'pendingDelete'
  };

  const tree = renderer
    .create(<RawComponent
      node={node}
      editor={editor}
      classes={classes}
      onDelete={onDelete} />, {
      createNodeMock: el => {
        if (el.type === 'img') {
          return {
            naturalWidth: 100,
            naturalHeight: 100
          }
        }
      }
    })
    .toJSON();
  expect(tree).toMatchSnapshot();
});