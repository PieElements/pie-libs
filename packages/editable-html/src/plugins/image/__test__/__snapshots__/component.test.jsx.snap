// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className=""
  onFocus={undefined}
>
  <div
    aria-valuenow={50}
    className="MuiLinearProgress-root-1 MuiLinearProgress-primaryColor-2"
  >
    <div
      className="MuiLinearProgress-bar-8 MuiLinearProgress-primaryColorBar-3 MuiLinearProgress-determinateBar1-15"
      style={
        Object {
          "transform": "scaleX(0.5)",
        }
      }
    />
  </div>
  <img
    src={undefined}
    style={
      Object {
        "height": "50px",
        "width": "50px",
      }
    }
  />
</div>
`;
